src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_sslstream.c

```diff
@@ -101,314 +106,691 @@ static jobject ensureRemaining(JNIEnv* env, SSLStream* sslStream, jobject oldBuf
+        *read = rem;
+        ret = SSLStreamStatus_OK;
     }
     else
     {
-        (*env)->DeleteLocalRef(env, (*env)->CallObjectMethod(env, sslStream->appInBuffer, g_ByteBufferCompactMethod));
-        doUnwrap(env, sslStream);
-        AssertOnJNIExceptions(env);
-        return AndroidCryptoNative_SSLStreamRead(sslStream, buffer, offset, length);
+        ret = SSLStreamStatus_NeedData;
     }
+
+cleanup:
+    ReleaseLRef(env, data);
+    return ret;
 }
 
-void AndroidCryptoNative_SSLStreamWrite(SSLStream* sslStream, uint8_t* buffer, int offset, int length)
+PAL_SSLStreamStatus AndroidCryptoNative_SSLStreamWrite(SSLStream* sslStream, uint8_t* buffer, int length)
 {
-    /*
-        appOutBuffer.put(message);
-        doWrap();
-    */
+    assert(sslStream != NULL);
 
     JNIEnv* env = GetJNIEnv();
+    PAL_SSLStreamStatus ret = SSLStreamStatus_Error;
+
+    // byte[] data = new byte[] { <buffer> }
+    // appOutBuffer.put(data);
     jbyteArray data = (*env)->NewByteArray(env, length);
-    (*env)->SetByteArrayRegion(env, data, 0, length, (jbyte*)(buffer + offset));
-    (*env)->DeleteLocalRef(env, (*env)->CallObjectMethod(env, sslStream->appOutBuffer, g_ByteBufferPut2Method, data));
+    (*env)->SetByteArrayRegion(env, data, 0, length, (jbyte*)buffer);
+    IGNORE_RETURN((*env)->CallObjectMethod(env, sslStream->appOutBuffer, g_ByteBufferPutByteArray, data));
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+    int handshakeStatus;
+    ret = DoWrap(env, sslStream, &handshakeStatus);
+    if (ret == SSLStreamStatus_OK && IsHandshaking(handshakeStatus))
+    {
+        ret = SSLStreamStatus_Renegotiate;
+    }
+
+cleanup:
     (*env)->DeleteLocalRef(env, data);
-    doWrap(env, sslStream);
-    AssertOnJNIExceptions(env);
+    return ret;
 }
 
 void AndroidCryptoNative_SSLStreamRelease(SSLStream* sslStream)
 {
+    if (sslStream == NULL)
+        return;
+
     JNIEnv* env = GetJNIEnv();
-    ReleaseGRef(env, sslStream->sslContext);
-    ReleaseGRef(env, sslStream->sslEngine);
-    ReleaseGRef(env, sslStream->sslSession);
-    ReleaseGRef(env, sslStream->appOutBuffer);
-    ReleaseGRef(env, sslStream->netOutBuffer);
-    ReleaseGRef(env, sslStream->netInBuffer);
-    ReleaseGRef(env, sslStream->appInBuffer);
-    free(sslStream);
-    AssertOnJNIExceptions(env);
+    FreeSSLStream(env, sslStream);
+}
+
+int32_t AndroidCryptoNative_SSLStreamGetApplicationProtocol(SSLStream* sslStream, uint8_t* out, int* outLen)
+{
+    assert(sslStream != NULL);
+
+    JNIEnv* env = GetJNIEnv();
+
+    int32_t ret = FAIL;
+    INIT_LOCALS(loc, protocol, bytes);
+
+    // String protocol = sslEngine.getApplicationProtocol();
+    loc[protocol] = (*env)->CallObjectMethod(env, sslStream->sslEngine, g_SSLEngineGetApplicationProtocol);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    if (loc[protocol] == NULL)
+        goto cleanup;
+
+    // byte[] bytes = protocol.getBytes();
+    loc[bytes] = (*env)->CallObjectMethod(env, loc[protocol], g_StringGetBytes);
```

Comment:
Any particular reason we're using `getBytes` here instead of one of the built-in JNI methods for getting string data?